[iOS][CV2] Fix CollectionView2 on iOS iPad jumps/scrolls randomly when item size changes

devanathan-vaithiyanathan opened the pull request:
<!-- Please let the below note in for people that find this PR -->
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

<!--
!!!!!!! MAIN IS THE ONLY ACTIVE BRANCH. MAKE SURE THIS PR IS TARGETING MAIN. !!!!!!! 
-->
### Issue Details
In CollectionView2, when an item’s size changes, the CollectionView jumps or scrolls unexpectedly instead of keeping its current position.

### Description of Change

<!-- Enter description of the fix in this section -->
Replaced the cell update logic from invalidateLayout to reconfigureItems for the affected cell.
reconfigureItems (iOS 15+) is lighter and works correctly on iPad iOS 26.

### Issues Fixed

<!-- Please make sure that there is a bug logged for the issue being fixed. The bug should describe the problem and how to reproduce it. -->

Fixes #32796 
Fixes #33833 

<!--
Are you targeting main? All PRs should target the main branch unless otherwise noted.
-->

### Regarding test case
Since the issue reproduces only on iPadOS 26, I haven’t added a test case.

**Tested the behavior in the following platforms.**
- [ ] Android
- [ ] Windows
- [x] iOS
- [ ] Mac

| Before  | After  |
|---------|--------|
| **iOS**<br> <video src="https://github.com/user-attachments/assets/6b3d1860-fa1d-4085-a0f9-e57f48eea1ff" width="400" height="600"> | **iOS**<br> <video src="https://github.com/user-attachments/assets/89036aa6-010c-4c64-b4de-e7a503579c6f" width="400" height="600"> |

devanathan-vaithiyanathan committed:
Fix added

devanathan-vaithiyanathan committed:
Update ItemsViewController2.cs

devanathan-vaithiyanathan committed:
Update ItemsViewController2.cs

devanathan-vaithiyanathan committed:
Update ItemsViewController2.cs

devanathan-vaithiyanathan committed:
fix modified for #33833

rmarinho commented:
<!-- AI Summary -->

## 🤖 AI Summary

<!-- SECTION:PR-REVIEW -->
<details>
<summary>📊 <strong>Expand Full Review</strong></summary>

---

<details>
<summary><strong>🔍 Pre-Flight — Context & Validation</strong></summary>

---

<details>
<summary>📝 <strong>Review Session</strong> — <strong>fix modified for #33833</strong> · <a href="https://github.com/dotnet/maui/commit/a88fe75b07706421fac23556b8f22ce1d363c1c7"><code>a88fe75</code></a></summary>

---

**Issue:** #32796 - CollectionView2 on iOS iPad jumps/scrolls randomly when item size changes (iOS 26.0)
**Issue (2):** #33833 - [iOS] Grouped CollectionView scroll jump/skip during normal scrolling (iOS 18.7.3)
**PR:** #32874 - [iOS][CV2] Fix CollectionView2 on iOS iPad jumps/scrolls randomly when item size changes
**Platforms Affected:** iOS / iPadOS (iPad-specific reproduction, also affects MacCatalyst due to .ios.cs file scope)
**Files Changed:** 1 implementation file (no test files)

### Issue Summary
In `CollectionView2`, when an item's size changes at runtime (e.g., expand/collapse), the CollectionView unexpectedly jumps or scrolls to a different position instead of maintaining the current scroll offset. Reproducible on iPad/iPadOS 18+ / iOS 26, NOT reproducible on iPhone. Two separate issues with the same root cause.

### Fix Approach (PR)
In `ItemsViewController2.InvalidateLayoutIfItemsMeasureChanged()`, instead of calling `InvalidateLayout` on the layout context for cells with changed sizes, the PR introduces:
- A new `ShouldApplyCellReConfiguration()` static method that returns `true` for iOS 15+
- When true: calls `collectionView.ReconfigureItems(...)` wrapped in `UIView.PerformWithoutAnimation`
- This uses the iOS 15+ API specifically designed for cell content updates without recreation, avoiding scroll position disruption

### Reviewer Comments (from copilot-pull-request-reviewer)
| File:Line | Reviewer Says | Author Says | Status |
|-----------|---------------|-------------|--------|
| ItemsViewController2.cs:248 | `ShouldApplyCellReConfiguration()` should also check `OperatingSystem.IsMacCatalystVersionAtLeast(15)` since .ios.cs compiles for both platforms | N/A | ⚠️ NEEDS FIX |
| ItemsViewController2.cs:227 | Comment says "Apple bug in iPadOS 18+" but no bug reference (radar/forum) | N/A | ⚠️ NEEDS UPDATE |
| ItemsViewController2.cs:226 | Comment says "iPadOS 18+" but linked issue #32796 says "iOS 26.0" (same version, different naming) | N/A | ⚠️ NEEDS CLARIFICATION |

### Code Issues Identified
1. **Missing MacCatalyst version check**: `ShouldApplyCellReConfiguration()` returns `OperatingSystem.IsIOSVersionAtLeast(15)` only. Since `ItemsViewController2.cs` has a `.ios.cs` extension (compiles for both iOS and MacCatalyst), it must also check `OperatingSystem.IsMacCatalystVersionAtLeast(15)`. The existing pattern at line 165-171 of the same file demonstrates the correct approach: `OperatingSystem.IsIOSVersionAtLeast(11) || OperatingSystem.IsMacCatalystVersionAtLeast(11)`.
2. **Version comment inconsistency**: "iPadOS 18+" and "iOS 26.0" refer to the same OS (Apple rebranded iOS 18 as iOS 26). Should be clarified.

### Fix Candidates
| # | Source | Approach | Test Result | Files Changed | Notes |
|---|--------|----------|-------------|---------------|-------|
| PR | PR #32874 | Use `ReconfigureItems` (iOS 15+) instead of `InvalidateLayout` for cells with changed sizes, wrapped in `PerformWithoutAnimation` | ⏳ PENDING (Gate) | `ItemsViewController2.cs` (+19) | Missing MacCatalyst check in `ShouldApplyCellReConfiguration()` |

</details>

</details>

---

<details>
<summary><strong>🚦 Gate — Test Verification</strong></summary>

---

<details>
<summary>📝 <strong>Review Session</strong> — <strong>fix modified for #33833</strong> · <a href="https://github.com/dotnet/maui/commit/a88fe75b07706421fac23556b8f22ce1d363c1c7"><code>a88fe75</code></a></summary>

---

**Result:** ❌ FAILED - No tests exist
**Platform:** ios
**Mode:** N/A - Tests not present

### Findings
- No test files found for issue #32796 or #33833 in `TestCases.HostApp` or `TestCases.Shared.Tests`
- PR author explicitly stated: "Since the issue reproduces only on iPadOS 26, I haven't added a test case."
- The issue involves visual scroll position behavior during dynamic item size changes, which is difficult to automate but not impossible

### Impact
Gate FAILED. Cannot verify that automated tests catch the regression. The PR has no automated test coverage.

</details>

</details>

---

<details>
<summary><strong>🔧 Fix — Analysis & Comparison</strong></summary>

---

<details>
<summary>📝 <strong>Review Session</strong> — <strong>fix modified for #33833</strong> · <a href="https://github.com/dotnet/maui/commit/a88fe75b07706421fac23556b8f22ce1d363c1c7"><code>a88fe75</code></a></summary>

---

**Skipped:** Gate FAILED (no tests found), Phase 3 was not executed per workflow rules.

**Exhausted:** N/A (skipped)
**Selected Fix:** PR's fix (Gate did not pass; no alternative exploration performed. The PR's fix was evaluated via code review only.)

</details>

</details>

---

<details>
<summary><strong>📋 Report — Final Recommendation</strong></summary>

---

<details>
<summary>📝 <strong>Review Session</strong> — <strong>fix modified for #33833</strong> · <a href="https://github.com/dotnet/maui/commit/a88fe75b07706421fac23556b8f22ce1d363c1c7"><code>a88fe75</code></a></summary>

---

## ⚠️ Final Recommendation: REQUEST CHANGES

### Summary

PR #32874 addresses a real and verified bug in CollectionView2 where item size changes cause unexpected scroll position jumps on iPad/iOS. The fix approach (using `ReconfigureItems` instead of `InvalidateLayout`) is conceptually sound and aligns with Apple's recommended API for this use case. However, the PR has two blocking issues that must be addressed before it can be merged.

### Root Cause

In `ItemsViewController2.InvalidateLayoutIfItemsMeasureChanged()`, when visible cells have changed sizes, the code calls `InvalidateLayout` on a `UICollectionViewLayoutInvalidationContext`. On iPadOS/iOS 18+ (a.k.a. iOS 26) with `UICollectionViewCompositionalLayout`, this causes UIKit to recalculate layout and disrupt the current scroll offset. The `ReconfigureItems` API (iOS 15+) is the correct replacement: it triggers a reconfigure callback on the cell without full layout invalidation, preserving scroll position.

### Fix Quality

The core fix is reasonable — `ReconfigureItems` wrapped in `PerformWithoutAnimation` is the right mechanism to update cell content/size while avoiding scroll jumps. However:

**Issue 1 (Blocking): Missing MacCatalyst version check**

`ShouldApplyCellReConfiguration()` currently reads:
```csharp
static bool ShouldApplyCellReConfiguration()
{
    return OperatingSystem.IsIOSVersionAtLeast(15);
}
```

Since `ItemsViewController2.cs` uses the `.ios.cs` extension, it compiles for **both iOS and MacCatalyst**. The `ReconfigureItems` API requires macOS Catalyst 15.0+. Without the MacCatalyst check, on MacCatalyst < 15 the code will either fail or fall through to the old path unexpectedly. The existing pattern at line 165-171 of the same file demonstrates the correct approach:

```csharp
static bool ShouldApplyCellReConfiguration()
{
    return OperatingSystem.IsIOSVersionAtLeast(15)
        || OperatingSystem.IsMacCatalystVersionAtLeast(15);
}
```

**Issue 2 (Blocking): No automated tests**

The PR author states the issue only reproduces on iPadOS 26. While visual scroll jump behavior is hard to test automatically, the underlying code path (`InvalidateLayoutIfItemsMeasureChanged`) can be unit tested or device-tested by simulating a cell size change event. Lack of tests means regressions could go undetected.

**Non-blocking observations:**
- The comment "Apple bug in iPadOS 18+" vs the issue saying "iOS 26.0" is potentially confusing (these are the same OS under different names). Suggest updating to "UICollectionViewCompositionalLayout layout invalidation issue on iPadOS 18+ (also known as iOS 26)" or similar.
- No Apple bug report reference is provided. If one exists, it should be included for future maintainers.
- The version gating is very broad: `IsIOSVersionAtLeast(15)` means ALL iOS 15+ users get the new code path. This might be intentional (the new path is strictly better), but it would be worth confirming the author tested on both newer and older iOS versions.

### Required Changes

1. **Fix `ShouldApplyCellReConfiguration()`** to include the MacCatalyst check:
   ```csharp
   static bool ShouldApplyCellReConfiguration()
   {
       return OperatingSystem.IsIOSVersionAtLeast(15)
           || OperatingSystem.IsMacCatalystVersionAtLeast(15);
   }
   ```

2. **Add automated tests** - even a basic UI test that changes an item's bound height and verifies the CollectionView scroll position remains stable would provide regression coverage. The `write-tests-agent` can help.

3. **Update comments** to resolve the "iPadOS 18+" vs "iOS 26" naming confusion and remove the "Apple bug" framing if no bug report number is available.

</details>

</details>

---

</details>
<!-- /SECTION:PR-REVIEW -->

<!-- SECTION:PR-FINALIZE -->
<details>
<summary>📋 <strong>Expand PR Finalization Review</strong></summary>

---

<details>

<summary><b>Title: ✅ Good</b></summary>

<br>

**Current:** `[iOS][CV2] Fix CollectionView2 on iOS iPad jumps/scrolls randomly when item size changes`
</details>

<details>

<summary><b>Description: ✅ Good</b></summary>

<br>

Description needs updates. See details below.
### ✨ Suggested PR Description

<!-- Please let the below note in for people that find this PR -->
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from this PR and let us know in a comment if this change resolves your issue. Thank you!

### Root Cause

In `CollectionView2` (`ItemsViewController2`), when a cell's measured size changes, `InvalidateLayoutIfItemsMeasureChanged()` is called to update the layout. The existing implementation calls `UICollectionViewLayout.InvalidateLayout(context)` with a `UICollectionViewLayoutInvalidationContext` targeting the affected cells.

On iPad with `UICollectionViewCompositionalLayout` and self-sizing cells, this layout invalidation causes the `UICollectionView` to reset the scroll offset as part of the re-layout cycle, producing a visible scroll jump to an unintended position. This was reported on iPadOS 18+ (iOS 18.7.3, labeled also as iOS 26 beta).

### Description of Change

**File:** `src/Controls/src/Core/Handlers/Items2/iOS/ItemsViewController2.cs`

In `InvalidateLayoutIfItemsMeasureChanged()`, when there are cells whose measured size has changed:

- **Before:** Called `collectionView.CollectionViewLayout.InvalidateLayout(context)` with a `UICollectionViewLayoutInvalidationContext` targeting affected index paths. On iPad with CompositionalLayout this triggers a scroll offset reset.
- **After:** On iOS 15+ / MacCatalyst 15+, calls `collectionView.ReconfigureItems(indexPaths)` inside `UIView.PerformWithoutAnimation()`. `ReconfigureItems` (iOS 15+) updates cell content in-place without full dequeue/recreation and without triggering the scroll-position side-effect of `InvalidateLayout`. `PerformWithoutAnimation` prevents any visual animation during the reconfiguration.

The original `InvalidateLayout` path is preserved as fallback for iOS/MacCatalyst < 15.

### Issues Fixed

Fixes #32796
Fixes #33833

### Platforms Tested

- [ ] Android
- [ ] Windows
- [x] iOS (iPad — primary repro platform)
- [ ] Mac

> **Note:** The fix applies to all iOS/MacCatalyst 15+ devices. Testing was performed on iPad, which is the primary platform where the jump occurs. iPhone behavior with the new code path should also be verified before merge.
</details>

<details>

<summary><b>Code Review: ⚠️ Issues Found</b></summary>

<br>

# Code Review — PR #32874

**File:** `src/Controls/src/Core/Handlers/Items2/iOS/ItemsViewController2.cs`

---

## 🔴 Critical Issues

### 1. Missing MacCatalyst version check in `ShouldApplyCellReConfiguration()`

**File:** `src/Controls/src/Core/Handlers/Items2/iOS/ItemsViewController2.cs`, line ~248

**Problem:**
```csharp
static bool ShouldApplyCellReConfiguration()
{
    return OperatingSystem.IsIOSVersionAtLeast(15);
}
```

This file lives in `Handlers/Items2/iOS/` and compiles for **both iOS and MacCatalyst** (the `iOS/` subfolder convention in MAUI means code runs on MacCatalyst as well). The `ReconfigureItems` API was introduced on both iOS 15 and MacCatalyst 15. Without the MacCatalyst check, the version guard is incomplete for MacCatalyst builds.

Every other multi-platform version check in this same file includes both platforms. For example:
```csharp
// Line ~165 in the same file
if (!(OperatingSystem.IsIOSVersionAtLeast(11) || OperatingSystem.IsMacCatalystVersionAtLeast(11) ...))
```

**Recommendation:**
```csharp
static bool ShouldApplyCellReConfiguration()
{
    return OperatingSystem.IsIOSVersionAtLeast(15)
        || OperatingSystem.IsMacCatalystVersionAtLeast(15);
}
```

The Copilot reviewer already flagged this in a review comment.

---

## 🟡 Suggestions

### 2. Comment inaccuracy: "iPadOS 18+" vs "iOS 26"

**File:** Lines 226–228

```csharp
// Workaround for Apple bug in iPadOS 18+ with UICollectionViewCompositionalLayout
// Self-sizing cells cause scroll position jumps during invalidation
```

**Problem:**  
- Issue #32796 is labeled `version/iOS-26` (iOS 26 beta) and the reporter says "iOS 26.0".
- Issue #33833 reports the problem on "iOS 18.7.3".  
- The PR description says "iPad iOS 26".
- The in-code comment says "iPadOS 18+".

These are inconsistent. "iOS 26" in Apple's versioning is not the same as "iPadOS 18" — iOS 26 is the name for the upcoming major release (developer preview / beta 2025), while iPadOS 18 is the current released version. It's unclear whether this affects iOS 18 (18.7.3 per issue #33833) or iOS 26 beta only.

**Recommendation:** Clarify the affected versions based on actual testing. A neutral description like the Copilot reviewer suggested is safer:
```csharp
// Workaround for a layout issue with UICollectionViewCompositionalLayout on iPad
// where self-sizing cells cause scroll position jumps during invalidation.
// ReconfigureItems (iOS 15+) avoids this by updating cells in-place without
// triggering a scroll offset reset in the layout pass.
```

---

### 3. Fix is applied to all iOS 15+ devices, but the reported issue is iPad-specific

**Problem:**  
`ShouldApplyCellReConfiguration()` returns `true` for **all iOS 15+ devices**, including iPhone. However:
- Issue #32796 author: "I could not reproduce it on an iPhone yet. So for me it seems like it is only an issue on an iPad."
- Issue #33833: "This only reproduces on iPad iOS in my testing."

The PR entirely replaces the `InvalidateLayout` path with `ReconfigureItems` on iOS 15+. If `ReconfigureItems` has different behavior (e.g., does not trigger re-layout of auto-sizing cells on iPhone in some edge case), this could introduce regressions on iPhone.

**Recommendation:** At minimum, verify iPhone behavior is unchanged. If the behavior difference is only on iPad/CompositionalLayout, consider scoping the guard more narrowly (though a device model check at runtime is not typical MAUI practice). The author should document whether iPhone was tested.

---

### 4. Potential null IndexPath crash

**File:** Lines 232–234

```csharp
collectionView.ReconfigureItems(
    invalidatedCells.Select(CollectionView.IndexPathForCell).ToArray());
```

**Problem:**  
`UICollectionView.IndexPathForCell()` returns `null` when the cell is no longer part of the visible cells (e.g., it was removed from the collection view between when it was detected in the loop and when this line executes). If any `null` is in the resulting array, `ReconfigureItems` will likely throw a `NullReferenceException` or have undefined behavior.

Note: The same risk exists in the original `InvalidateItems` call below. However, `ReconfigureItems` is a newer API and its null-tolerance is less well-known.

**Recommendation:** Filter null index paths before passing to the API:
```csharp
collectionView.ReconfigureItems(
    invalidatedCells
        .Select(CollectionView.IndexPathForCell)
        .Where(ip => ip != null)
        .ToArray());
```

(The same null-filtering should ideally be applied to the existing `InvalidateItems` path as well.)

---

### 5. `PerformWithoutAnimation` may suppress expected animations

**File:** Lines 231–235

```csharp
UIView.PerformWithoutAnimation(() =>
{
    collectionView.ReconfigureItems(...);
});
```

**Problem:**  
`UIView.PerformWithoutAnimation` disables all Core Animation transactions within the block. This will suppress any animations tied to cell content updates. The original `InvalidateLayout` path had no such suppression.

This is likely intentional (to prevent the scroll-jump from being animated), but it's a behavioral change: any content updates inside cell views that previously animated (e.g., animated property changes triggered by `Bind`) will no longer animate on iOS 15+.

**Recommendation:** Add a comment explaining why `PerformWithoutAnimation` is necessary:
```csharp
// Wrap in PerformWithoutAnimation to prevent ReconfigureItems from animating
// the scroll position adjustment that would otherwise occur during the layout pass.
UIView.PerformWithoutAnimation(() =>
{
    collectionView.ReconfigureItems(...);
});
```

---

## ✅ Looks Good

- The overall approach of using `ReconfigureItems` instead of `InvalidateLayout` to avoid scroll-position side effects is sound. `ReconfigureItems` is the Apple-recommended API for updating cell content without full dequeue/recreation when cell data changes.
- The fallback to the original `InvalidateLayout` path for iOS/MacCatalyst < 15 is correct (even though in practice MAUI's minimum deployment target may be iOS 15+ already).
- The use of the helper method `ShouldApplyCellReConfiguration()` to centralize the version check is good practice for maintainability.
- The PR correctly targets `Items2/iOS/` (the active CollectionView2 handler for iOS/MacCatalyst), not the deprecated `Items/` handler.
</details>

---

</details>
<!-- /SECTION:PR-FINALIZE -->

devanathan-vaithiyanathan committed:
Update ItemsViewController2.cs

KarthikRajaKalaimani committed:
Fix CollectionView grid spacing updates for first row and column (#34527)

<!-- Please let the below note in for people that find this PR -->
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could [test the resulting
artifacts](https://github.com/dotnet/maui/wiki/Testing-PR-Builds) from
this PR and let us know in a comment if this change resolves your issue.
Thank you!

### Issue Details:

Horizontalspacing / Verticalspacing is not not applied to the first
column in GridItemLayout using CollectionView on Android platform.
        
### Root Cause:

The grid spacing was not being distributed symmetrically across the
active layout implementations, so edge items did not fully participate
when spacing changed at runtime.

### Description of Change:

- On Android, the fix in MauiRecyclerView.cs changes how RecyclerView
padding is handled for GridItemsLayout. Android was already using
SpacingItemDecoration, which applies half-spacing on all four sides of
each item. Previously, negative RecyclerView padding canceled that
spacing at the control edges. The branch keeps that negative-padding
behavior for non-grid layouts, but disables it for GridItemsLayout,
allowing the grid’s half-spacing to remain visible at the outer
perimeter. This makes the first row and first column visually respond
when spacing changes, but it also changes the grid behavior from spacing
only between items to spacing around the outside edges as well.

**Tested the behavior in the following platforms:**

- [x] Android
- [x] Windows
- [ ] iOS
- [ ] Mac

### Reference:

N/A

### Issues Fixed:

Fixes  #34257      

### Screenshots
| Before  | After  |
|---------|--------|
| <Video
src="https://github.com/user-attachments/assets/578dda69-1d60-474c-a6d8-23b3f9d29a50"
Width="300" Height="600"> | <Video
src="https://github.com/user-attachments/assets/7f3826e6-5922-4b6f-a6b9-de581b7db6c3"
Width="300" Height="600"> |

devanathan-vaithiyanathan committed:
Update ItemsViewController2.cs

kubaflo reviewed (changes requested):
Could you please check the ai's suggestions?

devanathan-vaithiyanathan committed:
Update ItemsViewController2.cs

devanathan-vaithiyanathan committed:
Update ItemsViewController2.cs

kubaflo reviewed (changes requested):
Could you please check the ai's suggestions?

kubaflo reviewed (changes requested):
Could you please check the ai's suggestions?

kubaflo reviewed (changes requested):
Could you please resolve conflicts?

devanathan-vaithiyanathan committed:
[iOS][CV2] Fix CollectionView2 on iOS iPad jumps/scrolls randomly when item size changes

Use ReconfigureItems (iOS 15+/MacCatalyst 15+) instead of InvalidateLayout for cells
with changed sizes, wrapped in PerformWithoutAnimation to prevent scroll position jumps
on iPadOS 18+ with UICollectionViewCompositionalLayout.

Fixes #32796
Fixes #33833

devanathan-vaithiyanathan committed:
Update ItemsViewController2.cs
